### DISCUSSION TITLE
PrimeFaces 10.0.0 DataExporter causing High CPU

### DISCUSSION THREAD

**Discussion by nsjulien** [primefaces/community#278] [PrimeFaces] [answered]
PrimeFaces DataExporter component causing High CPU consuming threads while exporting data from Datatable

PrimeFaces 10.0.0/JSF 2.2/WildFly 17/JDK 8/11/POI 4.1.1

**Comment by Rapster** [comment]
And did you investigate what could be the cause? Try with latest 13

**Reply by nsjulien** [reply]
Yes thread dump shows a Runnable thread with high CPU time in the export method in the DataTableExporter class

**Reply by melloware** [reply]
How many rows are you exporting and are you using XLS or XLSXSTREAM?  XLSXSTREAM is much faster and more memory efficient than XLS or XLSX versions of Excel.

**Reply by nsjulien** [reply]
XLSX type with no of rows exported between 5000 to 100,000

**Reply by melloware** [reply, marked as answer]
`XLSX` stores ALL the data in memory which is why you are getting 100% CPU with 100,000 rows.  Switch to `XLSXSTREAM`.  Your 100% is because the Garbage Collector is trying to recover memory but can't.

**Reply by nsjulien** [reply]
I am checking with XLSXSTREAM but this issue came after an upgrade from PF 6.2 to PF 10
Excel export having 10 to 15 columns and rows between 10,000 to 100,000 with autoSizeColumn=true in Excel Options causing 100% CPU

**Reply by melloware** [reply]
Yes AutoSizeColumn has to keep track of every cell to know how bug to make the columns.    You need to disable that for very large recordsets.

**Reply by itnoldev** [reply]
There is a way to use XLSXSTREAM and postprocess method? 
When I use XLSXSTREAM, the postprocess method can elaborate only 100 rows.

**Reply by melloware** [reply]
XLSXSTREAM should be possible it only keeps 3 rows in memory at a time so you might have to consult the POI docs for more info.  But we hand you the raw POI document to work with in PostProcess.

**Reply by itnoldev** [reply]
The problem is that when the postProcess method receives the POI document, the document has flushed all the rows except for the last 100 ones.

**Reply by melloware** [reply]
yeah unless there is a POI way of navgating back to those there is nothing PF can do at that point.  That is how XLSXSTREAM works.

**Reply by itnoldev** [reply]
I did it extending the DataTableExcelXStreamExporter class and overriding the postRowExport.
In the postRowExport method I can access to every row and then do my stuff.

**Reply by Rapster** [reply]
I introduced lately (in PF 13 ?) `DataExporter#onRowExport`, so you don't have to override anything
